feat(wren-ui):  Settings

andreashimin opened the pull request:
## Description
- Implement settings related APIs
- Integrate APIs in settings modal


## Tasks
- [x] Implement settings API
- [x] Implement reset project API
- [x] Implement update data source properties API
- [x] Integrate settings API to display information
- [x] Integrate update data sources properties API
- [x] Integrate reset API to project settings
- [x] Improve onboarding to update settings flows
### Additional tasks
- [x] According to the discussed conclusion, revert #138
- [x] Add PG data source to settings UI & API after it implemented

## Others
Need to do migration `yarn migrate`

wwwy3y3 commented on wren-ui/migrations/20240125081244_create_model_column_table.js:
```diff
@@ -7,6 +7,7 @@ exports.up = function (knex) {
     .createTable('model_column', (table) => {
       table.increments('id').comment('ID');
       table.integer('model_id').comment('Reference to model ID');
+      table.foreign('model_id').references('model.id').onDelete('CASCADE');
```
@andreashimin can you put this to a new migration file ? I think migration file will not rerun if content change.

Conceptually, I think we should treat migration files as immutable files.

andreashimin replied:
I have made new migration for adding foreign key according to the conclusion.

wwwy3y3 commented on wren-ui/migrations/20240126100753_create_metrics_measure_table.js:
```diff
@@ -10,17 +10,18 @@ exports.up = function (knex) {
   return knex.schema.createTable('metric_measure', (table) => {
     table.increments('id').comment('ID');
     table.integer('metric_id').comment('Reference to metric ID');
+    table.foreign('metric_id').references('metric.id').onDelete('CASCADE');
```
this too

wwwy3y3 commented on wren-ui/src/apollo/server/factories/onboardingFactory.ts:
```diff
@@ -297,51 +348,55 @@ export class DuckDBStrategy implements IDataSourceStrategy {
       catalog: 'wrenai',
       type: DataSourceName.DUCKDB,
-      initSql,
+      initSql: trim(initSql),
```
Just curious, why trim it ?

andreashimin replied:
Because the value may contain '/n SELECT xxx from xxx \n\n' etc 🤔️

wwwy3y3 replied:
ok noted.

wwwy3y3 commented on wren-ui/src/apollo/server/resolvers/projectResolver.ts:
```diff
@@ -38,6 +42,35 @@ export class ProjectResolver {
+      await ctx.modelService.deleteAllModelsByProjectId(project.id);
+
+      await ctx.projectService.deleteProject(project.id);
+    } catch {
+      // do nothing
+    }
```
a little bit weird that we ignore errors here

andreashimin replied:
It just needs to prevent the no project situation when doing reset,
I changed it back to this.
```ts
let project: Project;
try {
  project = await ctx.projectService.getCurrentProject();
} catch {
  // no project found
  return true;
}
```

wwwy3y3 commented on wren-ui/src/apollo/server/resolvers/projectResolver.ts:
```diff
@@ -131,10 +164,34 @@ export class ProjectResolver {
   ) {
     const { type, properties } = args.data;
-    await this.removeCurrentProject(ctx);
+    // Curretly only support one data source
+    await this.resetCurrentProject(_root, args, ctx);
```
What does it mean "Currently only support one data source" ?

andreashimin replied:
I changed it to `Currently only can create one project`, is that ok?

wwwy3y3 replied:
sure

fredalai commented on wren-ui/src/components/settings/DataSourceSettings.tsx:
```diff
@@ -1,26 +1,114 @@
+const SampleDatasetPanel = (props: Props) => {
+  const { sampleDataset } = props;
+  const templates = getTemplates();
+  const [startSampleDataset] = useStartSampleDatasetMutation({
```
redirect to the Home page when the user switches sample data.

fredalai reviewed (commented):
Hmm... Switching data source / sample data still encounters errors

![截圖 2024-04-19 下午5 44 42](https://github.com/Canner/WrenAI/assets/42527625/39a6b455-9344-49ec-943e-ef097369e6a5)

![截圖 2024-04-19 下午5 44 49](https://github.com/Canner/WrenAI/assets/42527625/e3aa7b25-b72b-4e2c-adcf-0573ec5e0b42)

andreashimin commented:
Hi @wwwy3y3, 
I have rebased the PG connector and also implemented to setting API. 
Please kindly help check again if I mis-changing the code. Thanks

@fredalai 
I have fixed the reviews, except nitpick needs to wait for @swwu91 to reply.

Another issue that you mentioned in the comment needs help from core-server members, and also relates to @onlyjackfrost 's PR #138, we can discuss it next Monday. Thanks

> Reproduce steps:
Select E-COMMERCE -> reset project -> Select BigQuery -> reset project -> Select NBA

andreashimin commented:
Hi @wwwy3y3, @fredalai I have done all the suggestions mentioned above.
Please kindly help review when you guys are available, thanks~

fredalai reviewed (approved):
LGTM for the frontend part.

wwwy3y3 reviewed (approved):
LGTM

onlyjackfrost commented on wren-ui/src/apollo/server/connectors/duckdbConnector.ts:
```diff
@@ -41,9 +41,6 @@ export class DuckDBConnector
     const { initSql, sessionProps } = prepareOptions;
-
-    // clear previous session props to prevent, invalid session props error from prior saveDataSource action.
-    await this.wrenEngineAdaptor.putSessionProps([]);
```
Thanks!

onlyjackfrost reviewed (approved):
LGTM
